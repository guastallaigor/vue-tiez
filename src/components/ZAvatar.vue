<template>
  <span class="z-avatar" :class="[size, {'dark': dark, 'stacked': stacked}]">
    <img :alt="alt" :title="alt" :src="src" />
  </span>
</template>

<script>
export default {
  name: 'z-avatar',
  props: {
    src: {
      type: String,
      required: true
    },
    alt: {
      type: String,
      default: ''
    },
    size: {
      type: String,
      default: '',
      validator: (v) => (['small', '', 'medium', 'large'].includes(v))
    },
    dark: {
      type: Boolean,
      default: false
    },
    stacked: {
      type: Boolean,
      default: false
    },
    show: {
      type: Number,
      default: 0
    }
  }
}
</script>

<style lang="scss">
@mixin size($width: 20px, $height: 20px) {
  width: $width;
  height: $height;
}

.z-avatar {
  &.small {
    @include size();
  }

  @include size(30px, 30px);

  &.medium {
    @include size(60px, 60px);
  }

  &.large {
    @include size(90px, 90px);
  }

  img {
    width: 100%;
    height: 100%;
  }

  display: inline-flex;
  align-items: center;
  line-height: 0;
  vertical-align: top;
  border-radius: 100%;
  overflow: hidden;
  border-width: 1px;
  border-style: solid;
  border-color: rgb(238, 238, 238);
  border-image: initial;
  background: rgb(255, 255, 255);
  -webkit-mask-image: -webkit-radial-gradient(center, circle cover, white, black);

  &.dark {
    border-color: rgb(51, 51, 51);
    background: rgb(0, 0, 0);
  }

  &.stacked {
    margin-left: -10px;
  }
}
</style>
