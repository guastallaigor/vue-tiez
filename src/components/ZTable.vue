<template>
  <table class="z-table" :class="{ dark }" :style="`width:${width}`">
    <thead v-if="hasItems">
      <tr>
        <th v-for="(header, keyH) in headers" :key="keyH">
          <div> {{ header.name }}</div>
        </th>
    </tr>
    </thead>
    <tbody v-if="hasItems">
      <tr v-for="(item, keyV) in items" :key="keyV" aria-roledescription="row">
        <td v-for="(header, keyI) in headers" :key="keyI">
          <div v-if="header.image && item.image">
            <z-avatar :src="item.image" :dark="dark" />
          </div>
          <div v-else>{{ item[header.key] }}</div>
        </td>
      </tr>
    </tbody>
    <slot v-else />
  </table>
</template>

<script>
export default {
  name: 'z-table',
  components: {
    ZAvatar: () => import('./ZAvatar')
  },
  props: {
    headers: {
      type: Array,
      default: () => ([])
    },
    items: {
      type: Array,
      default: () => ([])
    },
    width: {
      type: String,
      default: '100%'
    },
    dark: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    hasItems () {
      const { items } = this

      return items && items.length
    }
  }
}
</script>
