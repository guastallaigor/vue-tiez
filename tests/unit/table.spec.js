import { mount } from '@vue/test-utils'
import ZTable from '../../src/components/ZTable'
import ZAvatar from '../../src/components/ZAvatar'

describe('When I create the ZTable component', () => {
  const createComponent = (propsData = {}, slot) => {
    return mount(ZTable, {
      propsData: propsData,
      stubs: {
        ZAvatar
      },
      slots: {
        default: slot || 'ZTable'
      }
    })
  }

  const headers = [
    {
      name: '',
      key: 'img',
      image: true
    },
    {
      name: 'NAME',
      key: 'name'
    },
    {
      name: 'USERNAME',
      key: 'userName'
    }
  ]

  const items = [
    {
      image: 'https://zeit.co/api/www/avatar/?u=evilrabbit&s=160',
      name: 'Evil Rabbit',
      userName: 'evilrabbit'
    },
    {
      image: 'https://zeit.co/api/www/avatar/?u=rauchg&s=160',
      name: 'Guilermo Rauch',
      userName: 'rauchg'
    }
  ]

  it('should be a Vue instance', () => {
    const wrapper = createComponent()
    expect(wrapper.find(ZTable).isVueInstance()).toBe(true)
  })

  it('should have a default z-table state', () => {
    const wrapper = createComponent({ headers, items })
    const table = wrapper.find('.z-table')
    expect(table.exists()).toBe(true)
    expect(table.classes().length).toBe(1)
    expect(table.classes()).toEqual(['z-table'])
    expect(table.attributes().style).toBe('width: 100%;')
    const thead = wrapper.find('.z-table > thead')
    expect(thead.exists()).toBe(true)
    const ths = wrapper.findAll('.z-table > thead > tr > th')
    expect(ths.at(0).text()).toBe('')
    expect(ths.at(1).text()).toBe('NAME')
    expect(ths.at(2).text()).toBe('USERNAME')
    const tbody = wrapper.find('.z-table > tbody')
    expect(tbody.exists()).toBe(true)
    const tdsFirst = wrapper.findAll('.z-table > tbody > tr:first-child > td')
    expect(tdsFirst.at(0).html()).toBe('<td><div><span class="z-avatar"><img alt="" title="" src="https://zeit.co/api/www/avatar/?u=evilrabbit&amp;s=160"></span></div></td>')
    expect(tdsFirst.at(1).text()).toBe('Evil Rabbit')
    expect(tdsFirst.at(2).text()).toBe('evilrabbit')
    const tdsLast = wrapper.findAll('.z-table > tbody > tr:last-child > td')
    expect(tdsLast.at(0).html()).toBe('<td><div><span class="z-avatar"><img alt="" title="" src="https://zeit.co/api/www/avatar/?u=rauchg&amp;s=160"></span></div></td>')
    expect(tdsLast.at(1).text()).toBe('Guilermo Rauch')
    expect(tdsLast.at(2).text()).toBe('rauchg')
  })

  it('should have a different width', () => {
    const wrapper = createComponent({ headers, items, width: '50%' })
    const table = wrapper.find('.z-table')
    expect(table.exists()).toBe(true)
    expect(table.classes().length).toBe(1)
    expect(table.classes()).toEqual(['z-table'])
    expect(table.attributes().style).toBe('width: 50%;')
  })

  it('should have a dark class', () => {
    const wrapper = createComponent({ headers, items, dark: true })
    const table = wrapper.find('.z-table')
    expect(table.exists()).toBe(true)
    expect(table.classes().length).toBe(2)
    expect(table.classes()).toEqual(['z-table', 'dark'])
  })

  it('should have a table slot instead of headers and items', () => {
    const slot = '<thead><tr><th><div>Test with slot</div></th></tr></thead> <tbody><tr><td>Test with slot</td></tr></tbody>'
    const wrapper = createComponent({}, slot)
    const table = wrapper.find('.z-table')
    expect(table.exists()).toBe(true)
    expect(table.classes().length).toBe(1)
    expect(table.classes()).toEqual(['z-table'])
    const thead = wrapper.find('.z-table > thead')
    expect(thead.html()).toBe(slot.split('> <')[0] + '>')
    const tbody = wrapper.find('.z-table > tbody')
    expect(tbody.html()).toBe('<' + slot.split('> <')[1])
    expect(table.html()).toBe('<table class="z-table" style="width: 100%;"><!----> <thead><tr><th><div>Test with slot</div></th></tr></thead> <tbody><tr><td>Test with slot</td></tr></tbody></table>')
  })

  it('should mount the ZAvatar component', (done) => {
    const wrapper = mount(ZTable, {
      propsData: { headers, items }
    })
    const table = wrapper.find('.z-table')
    expect(table.exists()).toBe(true)
    expect(table.classes().length).toBe(1)
    expect(table.classes()).toEqual(['z-table'])
    setTimeout(() => {
      expect(wrapper.find(ZAvatar).exists()).toBe(true)
      done()
    })
  })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      headers,
      items,
      width: '100%',
      dark: false
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
