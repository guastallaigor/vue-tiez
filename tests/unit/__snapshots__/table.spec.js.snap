// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZTable component should match snapshot 1`] = `
<table class="z-table" style="width: 100%;">
  <thead>
    <tr>
      <th>
        <div> </div>
      </th>
      <th>
        <div> NAME</div>
      </th>
      <th>
        <div> USERNAME</div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr aria-roledescription="row">
      <td>
        <div><span class="z-avatar"><img alt="" title="" src="https://zeit.co/api/www/avatar/?u=evilrabbit&amp;s=160"></span></div>
      </td>
      <td>
        <div>Evil Rabbit</div>
      </td>
      <td>
        <div>evilrabbit</div>
      </td>
    </tr>
    <tr aria-roledescription="row">
      <td>
        <div><span class="z-avatar"><img alt="" title="" src="https://zeit.co/api/www/avatar/?u=rauchg&amp;s=160"></span></div>
      </td>
      <td>
        <div>Guilermo Rauch</div>
      </td>
      <td>
        <div>rauchg</div>
      </td>
    </tr>
  </tbody>
</table>
`;
